crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,789 @@
+
+                    if ipv6 != Ipv6Addr::UNSPECIFIED {
+                        builder.ip6(ipv6);
+                    }
+                    builder.udp6(ipv6_port);
+
+                    (IpMode::DualStack, (ipv6, ipv6_port).into())
+                }
+            };
+
+            // add fork id
+            let (chain, fork_id) = fork;
+            builder.add_value_rlp(chain, alloy_rlp::encode(fork_id).into());
+
+            // add other data
+            for (key, value) in other_enr_data {
+                builder.add_value_rlp(key, alloy_rlp::encode(value).into());
+            }
+
+            // enr v4 not to get confused with discv4, independent versioning enr and
+            // discovery
+            let enr = builder.build(sk).expect("should build enr v4");
+            let EnrCombinedKeyWrapper(enr) = enr.into();
+
+            trace!(target: "net::discv5",
+                ?enr,
+                "local ENR"
+            );
+
+            // backwards compatible enr
+            let bc_enr = NodeRecord::from_secret_key(socket, sk);
+
+            (enr, bc_enr, ip_mode, chain)
+        };
+
+        //
+        // 3. start discv5
+        //
+        let sk = discv5::enr::CombinedKey::secp256k1_from_bytes(&mut sk.secret_bytes()).unwrap();
+        let mut discv5 = match discv5::Discv5::new(enr, sk, discv5_config) {
+            Ok(discv5) => discv5,
+            Err(err) => return Err(Error::InitFailure(err)),
+        };
+        discv5.start().await.map_err(Error::Discv5Error)?;
+
+        // start discv5 updates stream
+        let discv5_updates = discv5.event_stream().await.map_err(Error::Discv5Error)?;
+
+        let discv5 = Arc::new(discv5);
+
+        //
+        // 4. add boot nodes
+        //
+        Self::bootstrap(bootstrap_nodes, &discv5)?;
+
+        let metrics = Discv5Metrics::default();
+
+        //
+        // 5. bg kbuckets maintenance
+        //
+        Self::spawn_populate_kbuckets_bg(lookup_interval, metrics.clone(), discv5.clone());
+
+        Ok((
+            Self { discv5, ip_mode, fork_id_key, discovered_peer_filter, metrics },
+            discv5_updates,
+            bc_enr,
+        ))
+    }
+
+    /// Bootstraps underlying [`discv5::Discv5`] node with configured peers.
+    fn bootstrap(
+        bootstrap_nodes: HashSet<BootNode>,
+        discv5: &Arc<discv5::Discv5>,
+    ) -> Result<(), Error> {
+        trace!(target: "net::discv5",
+            ?bootstrap_nodes,
+            "adding bootstrap nodes .."
+        );
+
+        let mut enr_requests = vec![];
+        for node in bootstrap_nodes {
+            match node {
+                BootNode::Enr(node) => {
+                    if let Err(err) = discv5.add_enr(node) {
+                        return Err(Error::Discv5ErrorStr(err))
+                    }
+                }
+                BootNode::Enode(enode) => enr_requests.push(task::spawn({
+                    let discv5 = discv5.clone();
+                    async move {
```

Comment:
I think all this does is just sending a message over a channel and await the response?
I don't think we need to spawn this as a task?

Reply:
it makes the requests run in parallel

Reply:
they get all sent to the same service.
doing this on separate tasks is unnecessary.